### PR TITLE
feat(storage): upgrade tokio to 1.38.0 and simplify memory limiter notify

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Previously due to the problem of tokio oneshot channel described in https://github.com/risingwavelabs/risingwave/issues/6617, in the memory limiter, when we notify waiters, we will have to store the senders to notify and then notify senders after release the lock to avoid deadlock.

With https://github.com/tokio-rs/tokio/pull/6558 gets merged and published in tokio 1.38.0, we don't need this workaround anymore. We can change to notify the sender within the lock, and on send error, we can release the quota without acquiring the lock to notify again. The following code snippet shows the changed logic.

```
fn may_notify_waiters(self: &Arc<Self>) {
    ...
    {
        // Previously have a vector to hold the tx before release the lock
        // let mut notify_waiters = vec![];

        let mut waiters = self.controller.lock();
        while let Some((_, quota)) = waiters.front() {
            ...
            let (tx, quota) = waiters.pop_front().unwrap();
            // Changed to release quota when returning with err
            if let Err(tracker) = tx.send(MemoryTracker::new(self.clone(), quota)) {
                tracker.release_quota_without_notify();
            }
        }
        drop(waiters);

        ...

        // Previously notify waiters after release the lock
        // for (tx, quota) in notify_waiters {
        //     let _ = tx.send(MemoryTracker::new(self.clone(), quota));
        // }
    }
}
```

The memory limiter is moved to a separate crate, so that we can write [loom](https://github.com/tokio-rs/loom) test on the memory limiter. A loom test `test_memory_limiter_drop` is added in this PR, which tests the scenario of future cancellation that can cause deadlock in previous tokio version. It's expected to fail in the previous tokio version 1.37.0, though it still gets passed. We may further investigate on why loom is not able to cover all the corner cases.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/hummock_memory_limiter/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    fn release_quota(&self, quota: u64) {
+        self.total_size.fetch_sub(quota, AtomicOrdering::SeqCst);
+    }
+
+    fn add_memory(&self, quota: u64) {
+        self.total_size.fetch_add(quota, AtomicOrdering::SeqCst);
+    }
+
+    fn may_notify_waiters(self: &Arc<Self>) {
+        // check `has_waiter` to avoid access lock every times drop `MemoryTracker`.
```
Comment (wenym1) [inline src/storage/hummock_memory_limiter/src/lib.rs:52]: Only this method is changed compared to the original code in `src/storage/hummock/utils.rs`. 

Changed logic is described as in the PR description.

**Review by xxchan — COMMENTED** [review]
Perhaps we should upgrade `tokio` in another PR first (and ideally run benchmarks/longevity tests first), because it might be one of the most critical dependency.

If you've read the CHANGELOG and are very confident it's a harmless trivial upgrade, that's also fine.

**Comment by BugenZhao** [comment]
> the problem of tokio oneshot channel described in #6617,

This doesn't seem to be the correct issue number?

**Comment by wenym1** [comment]
> > the problem of tokio oneshot channel described in #6617,
> 
> This doesn't seem to be the correct issue number?

In the bullet 9 of the Investigation section in the issue description, you mentioned a corner case that msg might be dropped in the oneshot sender side even if the send returns ok.

```
pub fn send(mut self, t: T) -> Result<(), T> {
    let inner = self.inner.take().unwrap();

    inner.value.with_mut(|ptr| unsafe {
        *ptr = Some(t);
    });

    if !inner.complete() {
        unsafe {
            return Err(inner.consume_value().unwrap());
        }
    }

-> receiver dropped here

    Ok(())
}
```

**Comment by BugenZhao** [comment]
> In the bullet 9 of the Investigation section in the issue description, you mentioned a corner case that msg might be dropped in the oneshot sender side even if the send returns ok.

Wow, have almost no impression of this. Still find it interesting when rereading it now. 😄

**Comment by wenym1** [comment]
> Perhaps we should upgrade `tokio` in another PR first (and ideally run benchmarks/longevity tests first), because it might be one of the most critical dependency.
> 
> If you've read the CHANGELOG and are very confident it's a harmless trivial upgrade, that's also fine.

Make sense to me, I can have a separate PR for it. btw, I didn't see a place to specify the tokio version in any of the `Cargo.toml`, and the tokio version seems to be implicitly deduced from our set of dependencies. 

Should we instead explicitly specify the tokio version somewhere in the `Cargo.toml`? Though it may have naming conflict with the current madsim-tokio if we directly name it tokio.

**Comment by xxchan** [comment]
> I didn't see a place to specify the tokio version in any of the Cargo.toml, and the tokio version seems to be implicitly deduced from our set of dependencies.

Yes, I also thought about this problem before. I think we do need to specify the version explicitly. (Perhaps we should use `=version` to avoid upgrade by mistake.)

It cannot be put in workspace `Cargo.toml`, because no crate directly use it and it won't have any effect. Perhaps we can specify it in `src/cmd_all/Cargo.toml`, or `src/common/Cargo.toml` (Actually any package is OK I guess...)
 
sth like
```toml
# why we need this: XXX
dummy-tokio-used-to-restrict-version = { version= "=1.38.0", package = "tokio" ...}
```

**Comment by wenym1** [comment]
> > I didn't see a place to specify the tokio version in any of the Cargo.toml, and the tokio version seems to be implicitly deduced from our set of dependencies.
> 
> Yes, I also thought about this problem before. I think we do need to specify the version explicitly. (Perhaps we should use `=version` to avoid upgrade by mistake.)
> 
> It cannot be put in workspace `Cargo.toml`, because no crate directly use it and it won't have any effect. Perhaps we can specify it in `src/cmd_all/Cargo.toml`, or `src/common/Cargo.toml` (Actually any package is OK I guess...)
> 
> sth like
> 
> ```toml
> # why we need this: XXX
> dummy-tokio-used-to-restrict-version = { version= "=1.38.0", package = "tokio" ...}
> ```

https://github.com/risingwavelabs/risingwave/pull/17536
